### PR TITLE
[ELF] --save-temps --lto-emit-asm: derive ELF/asm file names from bitcode file names

### PR DESCRIPTION
**Pull request by MaskRay**
Port COFF's https://reviews.llvm.org/D78221 and
https://reviews.llvm.org/D137217 to ELF. For the in-process ThinLTO
link, `ld.lld --save-temps a.o d/b.o -o out` will create
ELF relocatable files `out.lto.a.o`/`d/out.lto.b.o` instead of
`out1.lto.o`/`out2.lto.o`. Deriving the LTO-generated relocatable file
name from bitcode file names helps debugging.

The relocatable file name from the first regular LTO partition does not
change: `out.lto.o`. The second, if present due to `--lto-partition=`,
changes from `out1.lto.o` to `lto.1.o`.

For an archive member, e.g. `d/a.a(coll.o at 8)`,
the relocatable file is `d/out.lto.a.a(coll.o at 8).o`.

`--lto-emit-asm` file names are changed similarly. `--lto-emit-asm -o
out` now creates `out.lto.s` instead of `out`, therefore the
`--lto-emit-asm -o -` idiom no longer works. However, I think this new
behavior (which matches COFF) is better since keeping or removing
`--lto-emit-asm` will dump different files, instead of overwriting the
`-o` output file from an executable/shared object to an assembly file.

### CODE REVIEW

File: lld/ELF/LTO.cpp [hunk 1]
```diff
@@ -352,32 +357,49 @@ std::vector<InputFile *> BitcodeCompiler::compile() {
...
+  bool savePrelink = config->saveTempsArgs.contains("prelink");
   std::vector<InputFile *> ret;
-  for (unsigned i = 0; i != maxTasks; ++i)
-    if (!buf[i].empty())
-      ret.push_back(createObjFile(MemoryBufferRef(buf[i], "lto.tmp")));
+  const char *ext = config->ltoEmitAsm ? ".s" : ".o";
+  for (unsigned i = 0; i != maxTasks; ++i) {
+    StringRef bitcodeFilePath;
+    StringRef objBuf;
+    if (files[i]) {
```
Comment (teresajohnson) [inline lld/ELF/LTO.cpp:371]: A comment explaining the two different cases handled here would be good.

File: lld/test/ELF/common-archive-lookup.s [hunk 2]
```diff
@@ -61,10 +61,11 @@
...
 # RUN: llvm-readobj --syms 11 | FileCheck --check-prefix=NFC %s
 
-# RUN: ld.lld -o - main.o 4.a --fortran-common --lto-emit-asm | FileCheck --check-prefix=ASM %s
+# RUN: ld.lld -o out main.o 4.a --fortran-common --lto-emit-asm
+# RUN: FileCheck --check-prefix=ASM %s < out.lto.s
 
-# RUN: ld.lld -o - main.o  --start-lib 1.bc 2.bc --end-lib --fortran-common --lto-emit-asm | \
-# RUN:   FileCheck --check-prefix=ASM %s
+# RUN: ld.lld -o out main.o --start-lib 1.bc 2.bc --end-lib --fortran-common --lto-emit-asm
+# RUN: FileCheck --check-prefix=ASM %s < out.lto.s
```
Comment (teresajohnson) [inline lld/test/ELF/common-archive-lookup.s:68]: Should out.lto.s be removed to ensure we aren't rechecking the one produced by the prior ld.lld invocation?

**Review by teresajohnson — COMMENTED** [review]
Are there any tests checking that --save-temps on something like a.o results in a.out.lto.a.o? I am only seeing enumerated temp files in the tests.

File: lld/test/ELF/common-archive-lookup.s (hunk 2 above)
Comment (MaskRay) [reply lld/test/ELF/common-archive-lookup.s:68]: Thanks for the suggestion. Removed.

File: lld/ELF/LTO.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/LTO.cpp:373]: Done!

Comment (teresajohnson) [reply lld/ELF/LTO.cpp:373]: Thanks, those are helpful. Can you also add a comment before 373 on meaning of files[i] being null vs not - reading through the comments you added it sounds like files[i] is non-null when object file was found in the cache?

File: lld/ELF/LTO.h [hunk 3]
```diff
@@ -46,8 +46,9 @@ class BitcodeCompiler {
 
 private:
   std::unique_ptr<llvm::lto::LTO> ltoObj;
-  std::vector<SmallString<0>> buf;
+  SmallVector<std::pair<std::string, SmallString<0>>, 0> buf;
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply lld/ELF/LTO.h:49]: ping on this comment.

File: lld/test/ELF/lto/cache.ll [hunk 4]
```diff
@@ -53,10 +53,10 @@
 
 ; RUN: rm -fr cache && mkdir cache
 ; RUN: ld.lld --thinlto-cache-dir=cache --save-temps -o out b.bc a.bc -M | FileCheck %s --check-prefix=MAP
-; RUN: ls out1.lto.o a.bc.0.preopt.bc b.bc.0.preopt.bc
+; RUN: ls out.lto.a.o a.bc.0.preopt.bc b.bc.0.preopt.bc
 
-; MAP: llvmcache-{{.*}}:(.text)
-; MAP: llvmcache-{{.*}}:(.text)
+; MAP: out.lto.b.o:(.text)
```
Comment (teresajohnson) [inline lld/test/ELF/lto/cache.ll:58]: Confused about what changed in the latest commit to cause these lines to need updating - was this just a missed test update in the prior commit?

Comment (MaskRay) [reply lld/test/ELF/lto/cache.ll:58]: This is a new test. I noticed the `-Map` difference, so I pre-committed a test and changed this PR to demonstrate the difference.

File: lld/ELF/LTO.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/LTO.cpp:373]: Clarified the comment a bit

File: lld/ELF/LTO.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/LTO.h:50]: Added a comment

**Review by teresajohnson — APPROVED** [review]

**Comment by ormris** [comment]
We're seeing a test failure on our buildbot after this patch landed. Could you take a look?

https://lab.llvm.org/buildbot/#/builders/216/builds/33382/steps/7/logs/FAIL__lld__defsym_ll
```
# RUN: at line 4
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
# RUN: at line 5
z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so -defsym=bar2=bar3 -save-temps
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so' -defsym=bar2=bar3 -save-temps
# RUN: at line 6
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o'
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# RUN: at line 7
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe -d Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe' -d 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so'
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# RUN: at line 10
z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o'
# RUN: at line 11
z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
# RUN: at line 12
z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so -defsym=bar2=bar3 -save-temps
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so' -defsym=bar2=bar3 -save-temps
# RUN: at line 13
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o'
# .---command stderr------------
# | z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe: error: 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o': no such file or directory
# `-----------------------------
# error: command failed with exit status: 1
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# `-----------------------------
# error: command failed with exit status: 2
--
********************
```

**Comment by MaskRay** [comment]
> We're seeing a test failure on our buildbot after this patch landed. Could you take a look?
> 
> [lab.llvm.org/buildbot/#/builders/216/builds/33382/steps/7/logs/FAIL__lld__defsym_ll](https://lab.llvm.org/buildbot/#/builders/216/builds/33382/steps/7/logs/FAIL__lld__defsym_ll)
> 
> ```
> # RUN: at line 4
> z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
> # RUN: at line 5
> z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so -defsym=bar2=bar3 -save-temps
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so' -defsym=bar2=bar3 -save-temps
> # RUN: at line 6
> z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o'
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
> # RUN: at line 7
> z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe -d Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe' -d 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so'
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
> # RUN: at line 10
> z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o'
> # RUN: at line 11
> z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
> # RUN: at line 12
> z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so -defsym=bar2=bar3 -save-temps
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so' -defsym=bar2=bar3 -save-temps
> # RUN: at line 13
> z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o'
> # .---command stderr------------
> # | z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe: error: 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o': no such file or directory
> # `-----------------------------
> # error: command failed with exit status: 1
> # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
> # .---command stderr------------
> # | FileCheck error: '<stdin>' is empty.
> # | FileCheck command line:  z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
> # `-----------------------------
> # error: command failed with exit status: 2
> --
> ********************
> ```

We're seeing a test failure on our buildbot after this patch landed. Could you take a look?

[lab.llvm.org/buildbot/#/builders/216/builds/33382/steps/7/logs/FAIL__lld__defsym_ll](https://lab.llvm.org/buildbot/#/builders/216/builds/33382/steps/7/logs/FAIL__lld__defsym_ll)

```
# RUN: at line 4
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-as.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
# RUN: at line 5
z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so -defsym=bar2=bar3 -save-temps
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so' -defsym=bar2=bar3 -save-temps
# RUN: at line 6
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so.lto.o'
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# RUN: at line 7
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe -d Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-objdump.exe' -d 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.so'
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# RUN: at line 10
z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o'
# RUN: at line 11
z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe -module-summary Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\opt.exe' -module-summary 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto/Inputs/defsym-bar.ll' -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o'
# RUN: at line 12
z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o -shared -o Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so -defsym=bar2=bar3 -save-temps
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp.o' 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp1.o' -shared -o 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so' -defsym=bar2=bar3 -save-temps
# RUN: at line 13
z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe --symbols Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o | z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe' --symbols 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o'
# .---command stderr------------
# | z:\b\llvm-clang-x86_64-sie-win\build\bin\llvm-readelf.exe: error: 'Z:\b\llvm-clang-x86_64-sie-win\build\tools\lld\test\ELF\lto\Output\defsym.ll.tmp2.so1.lto.o': no such file or directory
# `-----------------------------
# error: command failed with exit status: 1
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe' --check-prefix=OBJ 'Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll'
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  z:\b\llvm-clang-x86_64-sie-win\build\bin\filecheck.exe --check-prefix=OBJ Z:\b\llvm-clang-x86_64-sie-win\llvm-project\lld\test\ELF\lto\defsym.ll
# `-----------------------------
# error: command failed with exit status: 2
--
********************
```

Sorry for the failure. I did not remove stale output files and failed to noticed this.
Should be fixed by 4fcd7cf22deff4a63d2bac12c909be7266f8b353 (more tests using a lot of temporary outputs should consider `RUN: rm -rf %t && mkdir %t && cd %t`)

**Pull request merged**